integration/setup/bastion_instance.go

```diff
@@ -0,0 +1,559 @@
+			// This here might happen in case the bastion instance got already
+			// deleted.
+		} else if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
+	{
+		err = deleteBastionSecurityGroup(ctx, clusterID, config)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
+	return nil
+}
+
+func deleteBastionSecurityGroup(ctx context.Context, clusterID string, config Config) error {
+	var err error
+
+	var bastionSecurityGroupID string
+	{
+		config.Logger.LogCtx(ctx, "level", "debug", "message", "finding bastion security group")
+
+		i := &ec2.DescribeSecurityGroupsInput{
+			Filters: []*ec2.Filter{
+				{
+					Name:   aws.String("tag:Name"),
+					Values: []*string{aws.String(clusterID + "-bastion")},
+				},
+				{
+					Name:   aws.String("tag:giantswarm.io/cluster"),
+					Values: []*string{aws.String(clusterID)},
+				},
+			},
+		}
+
+		o, err := config.AWSClient.EC2.DescribeSecurityGroups(i)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+		if len(o.SecurityGroups) != 1 {
+			return microerror.Maskf(executionFailedError, "expected one security group, got %d", len(o.SecurityGroups))
+		}
+
+		bastionSecurityGroupID = *o.SecurityGroups[0].GroupId
+
+		config.Logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("found bastion security group %#q", bastionSecurityGroupID))
+	}
+
+	{
+		config.Logger.LogCtx(ctx, "level", "debug", "message", "deleting bastion security group")
+
+		i := &ec2.DeleteSecurityGroupInput{
+			GroupId: aws.String(bastionSecurityGroupID),
+		}
+
+		_, err = config.AWSClient.EC2.DeleteSecurityGroup(i)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		config.Logger.LogCtx(ctx, "level", "debug", "message", "deleted bastion security group")
+	}
+
+	return nil
+}
+
+func terminateBastionInstance(ctx context.Context, clusterID string, config Config) error {
+	var err error
+
+	var instanceID string
+	{
+		config.Logger.LogCtx(ctx, "level", "debug", "message", "finding bastion instance id")
+
+		i := &ec2.DescribeInstancesInput{
+			Filters: []*ec2.Filter{
+				{
+					Name: aws.String("tag:giantswarm.io/cluster"),
+					Values: []*string{
+						aws.String(clusterID),
+					},
+				},
+				{
+					Name: aws.String("tag:giantswarm.io/instance"),
+					Values: []*string{
+						aws.String("e2e-bastion"),
+					},
+				},
+				{
+					Name: aws.String("instance-state-name"),
+					Values: []*string{
+						aws.String(ec2.InstanceStateNamePending),
+						aws.String(ec2.InstanceStateNameRunning),
+						aws.String(ec2.InstanceStateNameStopped),
+						aws.String(ec2.InstanceStateNameStopping),
+					},
+				},
+			},
+		}
+
+		o, err := config.AWSClient.EC2.DescribeInstances(i)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		if len(o.Reservations) == 0 {
+			return microerror.Maskf(notExistsError, "bastion instance")
+		}
+		if len(o.Reservations) != 1 {
+			return microerror.Maskf(executionFailedError, "expected one bastion instance, got %d", len(o.Reservations))
```

Comment:
If we get multiple EC2 instances for given cluster with "purpose" of being `e2e-bastion` instances, wouldn't it be ok to just terminate all of them? What could go wrong?

Reply:
Same as what I told Pawel. I copied the code from the tccp resource and I want it to be aligned. Reason is I want to be able to abstract it. Having snowflakes here and there will increase complexity and this code here works fine so far. I would then rather iterate and fix problems when we have them.